### PR TITLE
cmd/snap-confine: nvidia: preserve globbed file prefix

### PR DESCRIPTION
**Pull request by bboozzoo**
When symlinking libgl files from hostfs we discard the path prefix inside
source_dir of a globbed file and overwrite already symlinked files if those are
created before.

For example, nvidia globs:
*  `libnvidia-tls.so*`
*  `tls/libnvidia-tls.so*`

Produce the following matches:

*  `${source_dir}/libnvidia-tls.so`
*  `${source_dir}/libnvidia-tls.so.390.42`
*  `${source_dir}/tls/libnvidia-tls.so`
*  `${source_dir}/tls/libnvidia-tls.so.390.42`

When the files under `${source_dir}/tls` are processed, eg.
`${source_dir}/tls/libnvidia-tls.so.390.42`, the symlink we create will
overwrite `${libgl_dir}/libnvidia-tls.so.390.42`, thus shadowing the original
library that was present at this path.

The patch attempts to fix this behavior and mirrors the `source_dir` globs, thus
the following symlinks are created:

*  `${libgl_dir}/libnvidia-tls.so.390.42 -> ${source_dir}/libnvidia-tls.so.390.42`
*  `${libgl_dir}/tls/libnvidia-tls.so.390.42 -> ${source_dir}/tsls/libnvidia-tls.so.390.42`

### CODE REVIEW

File: cmd/snap-confine/mount-support-nvidia.c [hunk 1]
```diff
@@ -145,10 +148,27 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
 	for (size_t i = 0; i < glob_res.gl_pathc; ++i) {
 		char symlink_name[512] = { 0 };
 		char symlink_target[512] = { 0 };
+		char prefix_dir[512] = { 0 };
 		const char *pathname = glob_res.gl_pathv[i];
 		char *pathname_copy
 		    SC_CLEANUP(sc_cleanup_string) = strdup(pathname);
+		char *dirname_copy
+		    SC_CLEANUP(sc_cleanup_string) = strdup(pathname);
```
Comment (zyga) [inline cmd/snap-confine/mount-support-nvidia.c]: Please check for both of those being NULL.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 2]
```diff
@@ -145,10 +148,27 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
...
 		char *filename = basename(pathname_copy);
+		char *directory_name = dirname(dirname_copy);
+		strncpy(prefix_dir, libgl_dir, sizeof prefix_dir);
```
Comment (zyga) [inline cmd/snap-confine/mount-support-nvidia.c]: Can you please use `sc_must_snprintf` instead, it's a bit easier to get right than knowing exactly how strncpy fails.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 3]
```diff
@@ -145,10 +148,27 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
...
+
+		if (strlen(directory_name) > source_dir_len) {
```
Comment (zyga) [inline cmd/snap-confine/mount-support-nvidia.c]: Hmm, can you add a comment what is going on here? I'm a bit slow today.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 4]
```diff
@@ -123,6 +125,7 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
 						   const char *glob_list[],
 						   size_t glob_list_len)
 {
+	size_t source_dir_len = strlen(source_dir);
```
Comment (jdstrand) [inline cmd/snap-confine/mount-support-nvidia.c]: strlen() can be tricky with untrusted inputs, but I verified the callers of this function and all of the inputs are trusted.

**Comment by bboozzoo** [comment]
I'll amend & force push if you don't mind. This will make cherry-picking easier in case we need it.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 5]
```diff
@@ -145,10 +148,35 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
 	for (size_t i = 0; i < glob_res.gl_pathc; ++i) {
 		char symlink_name[512] = { 0 };
 		char symlink_target[512] = { 0 };
+		char prefix_dir[512] = { 0 };
 		const char *pathname = glob_res.gl_pathv[i];
 		char *pathname_copy
 		    SC_CLEANUP(sc_cleanup_string) = strdup(pathname);
+		char *dirname_copy
+		    SC_CLEANUP(sc_cleanup_string) = strdup(pathname);
```
Comment (zyga) [inline cmd/snap-confine/mount-support-nvidia.c]: Did you mean to strdup pathname again? It looks like a bug

Comment (bboozzoo) [reply cmd/snap-confine/mount-support-nvidia.c]: Yes I totally mean it. POSIX `dirname()` modifies its input argument.

Comment (zyga) [reply cmd/snap-confine/mount-support-nvidia.c]: Can you call those pathname_copy1 and pathname_copy2 and add a comment describing why we need them. A naive quick look (I agree this is right but just confusing) may mistake this for copy-paste error.

Comment (bboozzoo) [reply cmd/snap-confine/mount-support-nvidia.c]: Force pushed

**Pull request merged**